### DISCUSSION TITLE
How to cast using node-red?

### DISCUSSION THREAD

**Discussion by matheuscanela** [tsightler/ring-mqtt#830] [Support Q&A] [answered]
Hey folks!
I'm trying to automate using the CAST from node-red; however, when I attempt to add the cast, it doesn't work.

**The stream works fine using the following formats:**
1. Through the VLC app
2. If I go to the Google Cast plugin in the Home Assistant dashboard and ask it to play from the cameras - which I added as per the manual
3. Using the Home Assistant "call service" to play the camera via stream

I tried the cast and castv2 nodes from the palette, but I'm still getting the errors below:

**From Cast:**
The device tries to open the video, but nothing happens, and node-red returns "Not able to load media:Load failed Not able to load the media."

**From Cast v2:**
Similar to the above, but it returns the error "Error: Unknown command."

Also, the HA logs shows the errors below
- Failed to cast media http://192.168.4.10:8123/api/hls/*****05a43859c6eda18d539a95/master_playlist.m3u8?authSig=**** from internal_url (http://192.168.4.10:8123). Please make sure the URL is: Reachable from the cast device and either a publicly resolvable hostname or an IP address
- Failed to cast media rtsp://ring:ring@192.168.4.10:8554/****576fa69a_live. Please make sure the URL is: Reachable from the cast device and either a publicly resolvable hostname or an IP address
- Failed to cast media rtsp://192.168.4.10:8554/****576fa69a_live. Please make sure the URL is: Reachable from the cast device and either a publicly resolvable hostname or an IP address
- Failed to cast media rtsp://homeassistant:8554/*****576fa69a_live. Please make sure the URL is: Reachable from the cast device and either a publicly resolvable hostname or an IP address
- Failed to cast media http://192.168.4.10:8123/api/hls/*****cfc7e5550bdd2d64db450f2f820df2c4bcdca6d730d63f0f30/master_playlist.m3u8?authSig=***** from internal_url (http://192.168.4.10:8123). Please make sure the URL is: Reachable from the cast device and either a publicly resolvable hostname or an IP address
Error from stream worker: Error opening stream (INVALIDDATA, Invalid data found when processing input) rtsp://****:****@192.168.4.10:8554/****6fa69a_live
Error from stream worker: Timestamp discontinuity detected: last dts = 63243720, dts = 922937504

If I open the IP in VLC, it works normally. Also, all the devices are on the same network without any segregation. I have also tried using it without a password and with the hostname instead of the IP, but the issue persists.

The reason why I want to use the cast is that I want to set the live option and avoid the buffer settings.

Configuration

**cast**
![Screenshot 2024-05-05 015446](https://github.com/tsightler/ring-mqtt/assets/6635137/f159e2a9-23b3-436d-9867-d43ee9e29044)

**castv2**
![castv2](https://github.com/tsightler/ring-mqtt/assets/6635137/822e5983-eca6-4c9c-998a-19e608005606)
^^ the above image shows @192... but I removed the @ and it didn't work too.

**Comment by tsightler** [comment, marked as answer]
I'm far from an expert on node-red or cast devices, but I'm about 99% sure it isn't possible to directly cast RTSP.  As far as I know, Chromecast devices are limited to HLS, DASH, and maybe WebRTC (not 100% sure about that last one).

**Comment by matheuscanela** [comment]
Thanks @tsightler

I also posted it [here](https://github.com/i8beef/node-red-contrib-castv2/issues/103#issuecomment-2094387055), which talks about the same thing you mentioned.

I actually solved the problem by adding the go2rtc add-on and streaming it using [DashCast](https://flows.nodered.org/node/node-red-contrib-castv2), which is quite quick. Also, while researching, I found a ticket on the internet complaining that the Ring was starting every 10 minutes without stopping. I noticed the same issue, but I apparently fixed it by asking the CASTv2 to stop casting before reaching 10 minutes. Otherwise, I think the Google Nest continues to open the stream, even if it times out.

I can see that your add-on doesn't support two-way audio. Would you mind pointing me to where in the code I could potentially add this functionality? I may try to implement it once I finish some other projects. In my use case, I don't have Alexa devices, and Ring doesn't integrate with Google, so I am using Home Assistant, your add-on, and some logic to stitch them together. If two-way audio works, I will be able to talk to whoever rings my doorbell.

**Reply by tsightler** [reply]
The project uses ring-client-api which does support two-way audio and that part is easy enough to get working (you can see an example in Homebridge Ring plugin), but what I haven't been able to get past is getting everything proxied via go2rtc using the RTSP protocol.  To this point, this has been beyond my skill level, or, at least my patience, but I'll admit I haven't worked on it too much as it's just not an interesting use case for me personally.  Also, since Home Assistant itself doesn't really have proper support for two-way audio without using 3rd-party stuff like the WebRTC card, I've simply not gone all in trying to make it work.

Theoretically, I believe it should just be a matter of adding enough code to provide ONVIF-T support via RTSP and setting up a backchannel, however, today the RTSP output stream is generated by piping RTP streams received directly from via Ring over WebRTC directly into ffmpeg and letting it produce the RTSP output that is forwarded to go2rtc.  I believe this would need some significant re-work to support two-way audio, but I could be wrong.

Certainly would be open to accepting a PR that implemented two-way audio support in a non-disruptive way, even though it is somewhat outside of the goals of this project which is focused primarily on automation and not interactive use.  You might find something like Scrypted better targeted for your use case, I believe it already supports two-way audio because it has a native RTSP implementation.

**Reply by tsightler** [reply]
Oh, and code is all contained in `devices/camera-livestream.js`.

**Reply by tsightler** [reply]
One thing that is really interesting is that go2rtc 1.9.0, which was just released about a week ago, now supports backchannel audio via exec source, which is what ring-mqtt is using today.  I think this opens up the possibility of enabling two-way audio without completely changing how the code works today, but rather using a method similar to how homebridge-ring works.

Unfortunately, I just think it will be quite a while before I have enough time to sit and play with it, but if I'm right, this should make two-way audio more of a possibility.

**Reply by matheuscanela** [reply]
Nice! and thanks for sharing about Scripted. I am at the begging of my journey, so heaps of things to do. I will consider the option you mentioned above, plus scripted, as I am aiming to install cameras in the property, so Frigate was already part of the project.

I will keep you posted and let me know if you ended up adding it here.

**Reply by tsightler** [reply]
The more I think about this, I believe the real way to implement 2-way audio is probably to completely re-write the streaming pipeline.  Right now the flow is something like this:

`Ring---WebRTC--->werift---RTP--->ffmpeg---RTSP--->go2rtc---RTSP--->Client`

And, for users leveraging go2rtc + WebRTC/RTSPtoWebRTC then its:

`Ring---WebRTC--->werift---RTP--->ffmpeg---RTSP--->go2rtc---RTSP--->go2rtc---WebRTC-->client`

And, even more convoluted, when a client connects to the RTSP server, the go2rtc has to exec a command which sends a message to ring-mqtt to connect to Ring, start the WebRTC stream with werift and start the ffmpeg process to publish the stream.  Trying to add backchannel audio to that seems like there are a lot of places to go wrong as you'd have to negotiate a backchannel with another ffmpeg process, mix that into the return stream with WebRTC, etc.

However, go2rtc supports WebRTC directly as a source, including it's own websocket based control channel, and this fully supports and audio return channel.  The only issue is that go2rtc obviously doesn't know how to speak Ring's websocket protocol, however, ring-mqtt does know this, so it should be trivial to create a go2rtc to Ring protocol translator.

Then the path would be:

`Ring---WebRTC--->go2rtc---RTSP--->Client`

This would dramatically simplify the code, significantly lower CPU and memory requirements, likely lower latency and reduce startup time, and enable backchannel audio.  That should effectively get 2-way audio with no complex coding since it is already supported via WebRTC.

**Reply by matheuscanela** [reply]
Makes sense, I will let you know if I get bandwidth to move forward with this feature
